raptor/raptor.py

```diff
@@ -44,26 +47,60 @@ def __init__(self, app, binary):
         except NotImplementedError:
             self.profile = None
 
+        # Create the runner
+        cmdargs = []
+        if app == 'chrome':
+            cmdargs.append('--load-extension={}'.format(
+                           os.path.join(webext_dir, 'raptor-chrome')))
+
+        self.output_handler = OutputHandler()
+        process_args = {
+            'processOutputLine': [self.output_handler],
+        }
+        runner_cls = runners[app]
+        self.runner = runner_cls(
+            binary, cmdargs=cmdargs, profile=self.profile, process_args=process_args)
+
     def start_control_server(self):
         self.control_server = RaptorControlServer()
         self.control_server.start()
 
-    def run_test(self, test):
+    def run_test(self, test, timeout=None):
         gen_test_url(self.app, test)
         install_webext(self.app, self.profile)
-        start_browser(self.app, self.binary, self.profile)
+        self.runner.start()
+        first_time = int(time.time()) * 1000
+        proc = self.runner.process_handler
+        self.output_handler.proc = proc
 
-    def process_results(self):
+        try:
+            self.runner.wait(timeout)
+        finally:
+            try:
+                self.runner.check_for_crashes()
+            except NotImplementedError:  # not implemented for Chrome
+                pass
+
+        if self.runner.is_running():
+            self.log("Application timed out after {} seconds".format(timeout))
+            self.runner.stop()
+
+        proc.output.append(
+            "__startBeforeLaunchTimestamp%d__endBeforeLaunchTimestamp"
+            % first_time)
+        proc.output.append(
+            "__startAfterTerminationTimestamp%d__endAfterTerminationTimestamp"
+            % (int(time.time()) * 1000))
+
+        self._process_results(proc.output)
```

Comment:
Ok, so this is the issue with using a webext for a runner. You can't dump anything to the standard out terminal from inside a google chrome web ext. You can from within Firefox using window.dump() but that's not supported on Chrome (and no other way).

So that's why the raptor web extensions post results to the control server - instead of dumping to stdout and being picked up by this output handler (like talos does now).

So this would work in Firefox if I dumped results with window.dump() but not in Google Chrome.

Is there an alternative way that the browser process can be killed besides looking for a keyword in the browser pid output? I was hoping somehow using mozrunner would provide us a way to do this. Any ideas?

Reply:
We can go ahead and merge this - this is still way better to use mozrunner etc. and maybe can come up with the way to kill the browsers in a different bug.

Reply:
Is there any way that the control server output can be handled by the same browser output handler? Then we could watch for the __raptor__Shutdown message to come from the control server instead of browser output and then kill the browser. That would work for firefox and chrome if it's possible.

Reply:
Maybe the control server could call OutputHandler.__call__() with whatever data it gets. Then we'd have both stdout and control server data passing through the OutputHandler. The OutputHandler could then look for the shutdown message like normal.

Reply:
That would be a good solution, I'll try that first. If that doens't work then I guess having the control server wait for a signal from the web ext and then somehow have the control server get the browser pid and then kill it, would be the next option.